### PR TITLE
[POAE7-2634] JITLib Code Improvement & Refactor

### PR DESCRIPTION
**Pull request by bigPYJ1151**
### What changes were proposed in this pull request?

- Add createLiteral for literal constant creation.
- Refactor createLocalJITValue to support recursive calling.
- Remove std::function from createLocalJITValue.
- Improve createVariable for setting initial value.

### Why are the changes needed?

For nextgen

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UTs

### Which label does this PR belong to?

Infra

### CODE REVIEW

File: cider/exec/nextgen/jitlib/base/ValueTypes.h [hunk 1]
```diff
@@ -206,6 +206,29 @@ inline uint64_t getJITTypeSize(JITTypeTag type_tag) {
   }
   return 0;
 }
+
+template <typename T>
+inline std::any castConstant(JITTypeTag target_type, T value) {
```
Comment (chaojun-zhang) [inline cider/exec/nextgen/jitlib/base/ValueTypes.h:211]: name to 'castLiteral'?

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp [hunk 2]
```diff
@@ -119,7 +126,8 @@ llvm::Value* createConstantImpl(llvm::LLVMContext& context, std::any value) {
   }
 }
 
-JITValuePointer LLVMJITFunction::createConstant(JITTypeTag type_tag, std::any value) {
+JITValuePointer LLVMJITFunction::createLiteralImpl(JITTypeTag type_tag,
+                                                   const std::any& value) {
   llvm::Value* llvm_value = nullptr;
   switch (type_tag) {
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:132]: can we just `llvm_value = createConstantImpl<decltype(type_tag)>(getLLVMContext(), value);`

File: cider/exec/nextgen/operators/ColumnToRowNode.cpp [hunk 3]
```diff
@@ -96,8 +96,9 @@ void ColumnToRowTranslator::codegen(context::CodegenContext& context) {
   ExprPtrVector& inputs = exprs;
 
   // for row loop
-  auto index = func->createVariable(JITTypeTag::INT64, "index");
-  index = func->createConstant(JITTypeTag::INT64, 0l);
+  // prototype:void func(CodegenContext* context, ArrowArray* in, ArrowArray* out);
+  auto arrow_pointer = func->getArgument(1);
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/operators/ColumnToRowNode.cpp:100]: L99-L100 not used, please delete.

File: cider/exec/nextgen/jitlib/base/JITFunction.h [hunk 4]
```diff
@@ -68,15 +68,44 @@ class JITFunction {
...
+      auto init_jit_value = createLiteral(type_tag, init_val);
+      return createVariableImpl(type_tag, name, init_jit_value);
+    }
+  }
+
+  [[deprecated("Use createLiteral.")]] JITValuePointer createConstant(
+      JITTypeTag type_tag,
+      const std::any& value) {
+    return createLiteralImpl(type_tag, value);
+  }
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cider/exec/nextgen/jitlib/base/JITFunction.h:88]: Has been labeled as deprecated, we could remove it soon.

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp (hunk 2 above)
Comment (bigPYJ1151) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:132]: It's unfeasible, ```decltype(type_tag)``` will be deducted as ```enum cider::jitlib::JITTypeTag```.

**Review by xieqi — APPROVED** [review]

**Pull request merged**